examples/fluids/qfunctions/sgs_dd_model.h

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+/// @file
+/// Structs and helper functions for data-driven subgrid-stress modeling
+/// See 'Invariant data-driven subgrid stress modeling in the strain-rate eigenframe for large eddy simulation' 2022 and 'S-frame discrepancy
+/// correction models for data-informed Reynolds stress closure' 2022
+
+#ifndef sgs_dd_model_h
+#define sgs_dd_model_h
+
+#include <ceed.h>
+
+#include "newtonian_state.h"
+#include "utils.h"
+#include "utils_eigensolver_jacobi.h"
+
+typedef struct SGS_DD_ModelContext_ *SGS_DDModelContext;
+struct SGS_DD_ModelContext_ {
+  CeedInt    num_inputs, num_outputs;
+  CeedInt    num_layers;
+  CeedInt    num_neurons;
+  CeedScalar alpha;
+
+  struct {
+    size_t bias1, bias2;
+    size_t weight1, weight2;
+    size_t out_scaling;
+  } offsets;
+  size_t     total_bytes;
+  CeedScalar data[1];
+};
+
+CEED_QFUNCTION_HELPER void LeakyReLU(CeedScalar *x, const CeedScalar alpha, const CeedInt N) {
+  for (CeedInt i = 0; i < N; i++) {
+    if (x[i] < 0) x[i] *= alpha;
+  }
+}
+
+CEED_QFUNCTION_HELPER void DataDrivenInference(const CeedScalar *inputs, CeedScalar *outputs, SGS_DDModelContext sgsdd_ctx) {
+  const CeedInt     num_neurons = sgsdd_ctx->num_neurons;
+  const CeedInt     num_inputs  = sgsdd_ctx->num_inputs;
+  const CeedInt     num_outputs = sgsdd_ctx->num_outputs;
+  const CeedScalar  alpha       = sgsdd_ctx->alpha;
+  const CeedScalar *bias1       = &sgsdd_ctx->data[sgsdd_ctx->offsets.bias1];
+  const CeedScalar *bias2       = &sgsdd_ctx->data[sgsdd_ctx->offsets.bias2];
+  const CeedScalar *weight1     = &sgsdd_ctx->data[sgsdd_ctx->offsets.weight1];
+  const CeedScalar *weight2     = &sgsdd_ctx->data[sgsdd_ctx->offsets.weight2];
+  CeedScalar        V[num_neurons];
+  SetValueN(V, 0, num_neurons);
```

Comment:
This will fail for GPUs due to variable-length-array incompatabilities. I'm not sure what the "correct way" to allocate a work vector inside a QFunction is. I could obviously allocate something in the context and then access that, but that seems like an abuse of the context.

Reply:
You can't allocate private workspace in the context because it's shared on the GPU. Do we need this to be dynamic or can it be a static specialization?

Reply:
Gotcha. I'm not sure what static specialization is. It doesn't necessarily need to be dynamically allocated; we could just always have an array set to some compile time constant and error out if the number of neurons requested is greater than that constant.

Reply:
Like you did for the mass projections, for example.

Reply:
Note that `V` is only a work vector and doesn't need to be persistent.

> Like you did for the mass projections, for example.

Meaning like doing a `DMClone` and bringing a `Vec` in as a collocated array? I think doing a CeedVector would probably be easier, as we need this vector per quadrature point and can thus just create one from an element restriction or a custom strided vector creation. I guess I'd need to have that be an output for the Operator, as I need to write and read from it.

Reply:
This is a small workspace, like 20 "neuron" hidden layer, right? You can put the weights/biases in the context and just use a constant size stack array for the hidden layer workspace.

Reply:
Yeah, that's right. I could just set a constant size 20 neuron work array and that would be fine for this neural network.